src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/UploadDocumentsHandler.java

```diff
@@ -0,0 +1,141 @@
+import uk.gov.hmcts.reform.bailcaseapi.domain.service.DocumentReceiver;
+import uk.gov.hmcts.reform.bailcaseapi.domain.service.DocumentsAppender;
+
+@Component
+public class UploadDocumentsHandler implements PreSubmitCallbackHandler<BailCase> {
+
+    private final DocumentReceiver documentReceiver;
+    private final DocumentsAppender documentsAppender;
+    private static final String SUPPLIED_BY_APPLICANT = "Applicant";
+    private static final String SUPPLIED_BY_LEGAL_REPRESENTATIVE = "Legal Representative";
+    private static final String SUPPLIED_BY_HOME_OFFICE = "Home Office";
+
+    public UploadDocumentsHandler(
+        DocumentReceiver documentReceiver,
+        DocumentsAppender documentsAppender
+    ) {
+        this.documentReceiver = documentReceiver;
+        this.documentsAppender = documentsAppender;
+    }
+
+    public boolean canHandle(
+        PreSubmitCallbackStage callbackStage,
+        Callback<BailCase> callback
+    ) {
+        requireNonNull(callbackStage, "callbackStage must not be null");
+        requireNonNull(callback, "callback must not be null");
+
+        return callbackStage == PreSubmitCallbackStage.ABOUT_TO_SUBMIT
+               && callback.getEvent() == Event.UPLOAD_DOCUMENTS;
+    }
+
+    public PreSubmitCallbackResponse<BailCase> handle(
+        PreSubmitCallbackStage callbackStage,
+        Callback<BailCase> callback
+    ) {
+        if (!canHandle(callbackStage, callback)) {
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        final BailCase bailCase = callback.getCaseDetails().getCaseData();
+
+        Optional<List<IdValue<DocumentWithDescription>>> maybeDocument = bailCase.read(UPLOAD_DOCUMENTS);
+        Optional<String> maybeSuppliedBy = bailCase.read(UPLOAD_DOCUMENTS_SUPPLIED_BY);
+
+        String suppliedBy = maybeSuppliedBy
+            .orElse(isLegalRep(bailCase)
+                        ? "Legal Representative"
+                        : isHomeOffice(bailCase)
+                ? "Home Office"
+                : "");
+
+        if (maybeDocument.isPresent()) {
+            List<DocumentWithMetadata> document =
+                maybeDocument
+                    .orElseThrow(() -> new IllegalStateException("document is not present"))
+                    .stream()
+                    .map(IdValue::getValue)
+                    .map(doc -> documentReceiver.tryReceive(doc, DocumentTag.UPLOAD_DOCUMENT, suppliedBy))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toList());
+
+            BailCaseFieldDefinition appropriateCollectionForUserRole =
+                selectAppropriateDocumentsCollection(bailCase, suppliedBy);
+
+            Optional<List<IdValue<DocumentWithMetadata>>> maybeExistingDocuments =
+                bailCase.read(appropriateCollectionForUserRole);
+
+            final List<IdValue<DocumentWithMetadata>> existingHomeOfficeDocuments =
```

Comment:
please rename this variable, as it is not HODocuments. line 91 can be mabeExistingDocumentsCollection and this can be mabeExistingDocuments.

Reply:
Renamed